core/modules/replica/IngestServerConnection.cc

```diff
@@ -0,0 +1,520 @@
+                                               boost::asio::io_service& io_service)
+    :   _serviceProvider(serviceProvider),
+        _workerName(workerName),
+        _workerInfo(serviceProvider->config()->workerInfo(workerName)),
+        _socket(io_service),
+        _bufferPtr(make_shared<ProtocolBuffer>(
+            serviceProvider->config()->requestBufferSizeBytes()
+        )) {
+}
+
+
+IngestServerConnection::~IngestServerConnection() {
+    _closeFile();
+}
+
+
+void IngestServerConnection::beginProtocol() {
+    _receiveHandshake();
+}
+
+
+void IngestServerConnection::_receiveHandshake() {
+
+    LOGS(_log, LOG_LVL_DEBUG, context << __func__);
+
+    const size_t bytes = sizeof(uint32_t);
+
+    _bufferPtr->resize(bytes);
+
+    boost::asio::async_read(
+        _socket,
+        boost::asio::buffer(_bufferPtr->data(), bytes),
+        boost::asio::transfer_at_least(bytes),
+        boost::bind(
+            &IngestServerConnection::_handshakeReceived,
+            shared_from_this(),
+            boost::asio::placeholders::error,
+            boost::asio::placeholders::bytes_transferred
+        )
+    );
+}
+
+
+void IngestServerConnection::_handshakeReceived(boost::system::error_code const& ec,
+                                                size_t bytes_transferred) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context << __func__);
+
+    if (::isErrorCode(ec, __func__)) return;
+
+    // Now read the body of the request
+
+    ProtocolIngestHandshakeRequest request;
+    if (not ::readMessage(_socket, _bufferPtr, _bufferPtr->parseLength(), request)) return;
+
+    _transactionId   = request.transaction_id();
+    _table           = request.table();
+    _chunk           = request.chunk();
+    _isOverlap       = request.is_overlap();
+    _columnSeparator = request.column_separator() == ProtocolIngestHandshakeRequest::COMMA ? ',' : '\t';
+
+    // Check if a context of the request is valid
+    
+    try {
+        auto transactionInfo = _serviceProvider->databaseServices()->transaction(_transactionId);
+        if (transactionInfo.state != "STARTED") {
+            _illegalParameters("transaction is not active");
+            return;
+        }
+
+        // Get and validate a status of the database and the table
+
+        _databaseInfo = _serviceProvider->config()->databaseInfo(transactionInfo.database);
+        if (_databaseInfo.isPublished) {
+            throw invalid_argument("database '" + _databaseInfo.name + "' is already PUBLISHED");
+        }
+        _isPartitioned = _databaseInfo.partitionedTables.end() != find(
+            _databaseInfo.partitionedTables.begin(),
+            _databaseInfo.partitionedTables.end(),
+            _table);
+        if (not _isPartitioned) {
+            if (_databaseInfo.regularTables.end() != find(
+                _databaseInfo.regularTables.begin(),
+                _databaseInfo.regularTables.end(),
+                _table)) {
```

Comment:
is a wrapped `if` supposed to be indented 2x? Otherwise it blends in with the body of the `if`, which can be hard to read.